src/transformers/modeling_utils.py

```diff
@@ -1995,72 +1996,67 @@ def save_pretrained(
                 logger.info(
                     "Detected adapters on the model, saving the model in the PEFT format, only adapter weights will be saved."
                 )
-                state_dict = model_to_save.get_adapter_state_dict()
 
-                if save_peft_format:
-                    logger.info(
-                        "To match the expected format of the PEFT library, all keys of the state dict of adapters will be pre-pended with `base_model.model`."
-                    )
-                    peft_state_dict = {}
-                    for key, value in state_dict.items():
-                        peft_state_dict[f"base_model.model.{key}"] = value
-                    state_dict = peft_state_dict
+                total_adapters = list(self.peft_config.keys())
+
+                for adapter_name in total_adapters:
+                    adapter_state_dict = model_to_save.get_adapter_state_dict(adapter_name=adapter_name)
+
+                    if save_peft_format:
+                        logger.info(
+                            "To match the expected format of the PEFT library, all keys of the state dict of adapters will be pre-pended with `base_model.model`."
+                        )
+                        peft_state_dict = {}
+                        for key, value in adapter_state_dict.items():
+                            peft_state_dict[f"base_model.model.{key}"] = value
+                        adapter_state_dict = peft_state_dict.copy()
+                        # Free memory
+                        del peft_state_dict
+
+                    current_peft_config = self.peft_config[adapter_name]
+                    peft_multi_adapter_state_dict[adapter_name] = adapter_state_dict
+
+                    # the default adapter is always saved on the root directory
+                    if adapter_name != "default":
+                        current_peft_config.save_pretrained(os.path.join(save_directory, adapter_name))
+                    else:
+                        current_peft_config.save_pretrained(save_directory)
+
+                if len(peft_multi_adapter_state_dict.keys()) == 1:
+                    current_adapter = list(peft_multi_adapter_state_dict.keys())[0]
+                    state_dict = peft_multi_adapter_state_dict[current_adapter].copy()
+                    peft_multi_adapter_state_dict = None
```

Comment:
How about this change, which I think makes the intent a bit more obvious and avoid changing the type of `peft_multi_adapter_state_dict`:

```suggestion
                if len(peft_multi_adapter_state_dict.keys()) == 1:
                    current_adapter = list(peft_multi_adapter_state_dict.keys())[0]
                    state_dict = peft_multi_adapter_state_dict.pop(current_adapter).copy()
```

Not sure if the `.copy()` is needed?

Then, below change:

```python
- _peft_save_multi_adapter = _hf_peft_config_loaded and peft_multi_adapter_state_dict is not None
+ _peft_save_multi_adapter = _hf_peft_config_loaded and peft_multi_adapter_state_dict
```